lib/pmc_miller/reader.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require "pmc_miller/data"
+
+# PmcMiller
+module PmcMiller
+  ##
+  # Readers read in puppet-metrics-collector data
+  #
+  # Readers are associated with a specific puppet-metrics-collector directory.
+  # They can then assemble data for a specific key after they have been
+  # associated with the desired service.
+  class Reader
+    require "json"
+
+    attr_accessor :service
+
+    ##
+    # Creates a new reader that will use the given path as its data source.
+    #
+    # @param path [String] File system path to puppet-metrics-collector data
+    #
+    # @return [PmcMiller::Reader]
+    #
+    def initialize(path)
+      @path = path
+      @service = nil
+    end
+
+    ##
+    # Reads data associated with the given key for the assigned service of the
+    # reader.
+    #
+    # @param key [Symbol] Key to lookup data for
+    #
+    # @return [PmcMiller::Data] Data collection found
+    #
+    def read(key)
+      return unless %i[puppetserver puppetdb].include? @service
+
+      data = PmcMiller::Data.new
+      json_files.each do |f|
+        json_string = File.read(f)
+        json = JSON.parse(json_string)
+        data << PmcMiller::DataPoint.new(json["timestamp"], nested_hash_value(json, key.to_s))
+      end
+      data
+    end
+
+    private
+
+    ##
+    # Recursively search through object for key
+    #
+    # Copied from https://stackoverflow.com/a/8301752
+    #
+    # @param obj [Object] Object to recursively search
+    # @param key [Symbol, string] Key to search for
+    #
+    # @return First found value associated with key, nil if not found
```

Comment:
I don't love the "first found" aspect here.  Seems error prone.  Especially when many of the keys will be the same like "mean"...  And when of course the order of the keys in the json aren't guaranteed to remain the same.  But that could be a separate ticket.  probably a config file mapping short names to full hierarchy of the key.

Reply:
This has been reverted and to re-establish the key_map logic for only honoring the `:queue_depth` query.